src/Chainweb/Pact/PactService.hs

```diff
@@ -418,246 +410,177 @@ serviceRequests memPoolAccess reqQ = do
-
-              newTrans <- getBlockTxs bfState
-              if V.null newTrans then pure unchanged else do
-
-                pairs <- execTransactionsOnly miner newTrans pdbenv
-                  (Just txTimeLimit) `catch` handleTimeout
-
-                (oldPairsLength, oldFailsLength) <- liftIO $ (,)
-                  <$> Vec.length successes
-                  <*> Vec.length failures
-
-                newState <- foldM (splitResults successes failures) unchanged pairs
-
-                -- LOOP INVARIANT: gas must not increase
-                when (_bfGasLimit newState > _bfGasLimit bfState) $
-                  throwM $ MempoolFillFailure $ "Gas must not increase: " <> sshow (bfState,newState)
-
-                (newPairsLength, newFailsLength) <- liftIO $ (,)
-                  <$> Vec.length successes
-                  <*> Vec.length failures
-                let newSuccessCount = newPairsLength - oldPairsLength
-                let newFailCount = newFailsLength - oldFailsLength
-
-                -- LOOP INVARIANT: gas must decrease ...
-                if (_bfGasLimit newState < _bfGasLimit bfState)
-                    -- ... OR only non-zero failures were returned.
-                   || (newSuccessCount == 0  && newFailCount > 0)
-                    then go (incCount newState)
-                    else throwM $ MempoolFillFailure $ "Invariant failure: " <>
-                         sshow (bfState,newState,V.length newTrans
-                               ,newPairsLength,newFailsLength)
-
-    incCount :: BlockFill -> BlockFill
-    incCount b = over bfCount succ b
-
-    splitResults success fails (BlockFill g rks i) (t,r) = case r of
-      Right cr -> do
-        !rks' <- enforceUnique rks (requestKeyToTransactionHash $ P._crReqKey cr)
-        -- Decrement actual gas used from block limit
-        let !g' = g - fromIntegral (P._crGas cr)
-        liftIO $ Vec.push success (t, cr)
-        return $ BlockFill g' rks' i
-      Left f -> do
-        !rks' <- enforceUnique rks (gasPurchaseFailureHash f)
-        -- Gas buy failure adds failed request key to fail list only
-        liftIO $ Vec.push fails f
-        return $ BlockFill g rks' i
-
-    enforceUnique rks rk
-      | S.member rk rks =
-        throwM $ MempoolFillFailure $ "Duplicate transaction: " <> sshow rk
-      | otherwise = return $ S.insert rk rks
-
-    pHeight = _blockHeight $ _parentHeader parent
-    pHash = _blockHash $ _parentHeader parent
-
-    updateMempool = liftIO $ do
-      mpaProcessFork mpAccess $ _parentHeader parent
-      mpaSetLastHeader mpAccess $ _parentHeader parent
+        handleTimeout :: TxTimeout -> PactBlockM logger cas a
+        handleTimeout (TxTimeout h) = liftPactServiceM $ do
+          logError $ "timed out on " <> sshow h
+          liftIO $ mpaBadlistTx mpAccess (V.singleton h)
+          throwM (TxTimeout h)
+
+        !parentTime =
+          ParentCreationTime (_blockCreationTime $ _parentHeader latestHeader)
```

Comment:
at some point we should rename `_parentHeader` to `_getHeader` or something - I always doubletake until i'm reminded that it's not accessing the parent header, just the header object from `ParentHeader` the data type.

Reply:
Agreed.